Add Clippy

killercup opened the pull request:
Add [clippy](https://github.com/Manishearth/rust-clippy) and fix all of its warnings. This is

- adding the latest clippy 0.0.103 as an optional dependency with the `lint` feature
- improving a whole bunch of stuff in the various crates to make clippy real happy
- green with `#[deny(warnings)]`!
- adding a new entry in Travis' build matrix (which compiles quite fast as it is using `no-trans`)

killercup commented:
Currently, [these](https://gist.github.com/killercup/2f921016c158a051b01ebd3cd5bf4e25) warnings remain.

killercup commented on diesel/src/lib.rs:
```diff
@@ -1,9 +1,13 @@
 //! interactions. [A getting started guide](http://diesel.rs/guides/getting-started/) can be
 //! found on our website.
-#![deny(warnings, missing_debug_implementations, missing_copy_implementations)]
+#![deny(missing_debug_implementations, missing_copy_implementations)]
```
This change should be reverted before merging (but it is really when checking for warnings)

sgrif commented on clippy.toml:
```diff
@@ -0,0 +1,2 @@
+cyclomatic-complexity-threshold = 30
+doc-valid-idents = ["MiB", "GiB", "TiB", "PiB", "EiB", "SQLite", "PostgreSQL"]
```
What is `MiB` and friends here? Sizes? Where did they come from?

killercup replied:
Those are clippy's default values, copied from the docs. There are actually defined [here](https://github.com/Manishearth/rust-clippy/blob/c66e90303fe7cd0df6f3191d157e65f4d58b5de5/src/utils/conf.rs#L148-L161) and I just noticed they also included 'GitHub'. I guess we could just leave them and add diesel's own CamelCase-names-that-are-not-code when we see an invalid warning?

sgrif replied:
Seems good

sgrif commented on diesel/src/lib.rs:
```diff
@@ -1,9 +1,13 @@
 
+#![cfg_attr(feature = "dev", allow(unstable_features))]
+#![cfg_attr(feature = "dev", feature(plugin))]
+#![cfg_attr(feature = "dev", plugin(clippy(conf_file="../clippy.toml")))]
```
What do you think about adding clippy to `dev-dependencies` instead, and changing this to `cfg_attr(all(test, feature = "unstable")))`

killercup replied:
I created a new feature so we can control if/when CI includes clippy. Do you think we should just include it every time unstable is tested?

sgrif replied:
I would imagine we'd want it whenever possible since it's a lint pass. I
don't feel terribly strongly about it though

On Tue, Apr 26, 2016, 2:36 AM Pascal Hertleif <notifications@github.com>
wrote:

> In diesel/src/lib.rs
> <https://github.com/diesel-rs/diesel/pull/305#discussion_r61047346>:
>
> >  #![cfg_attr(feature = "unstable", feature(specialization))]
> >
> > +#![cfg_attr(feature = "dev", allow(unstable_features))]
> > +#![cfg_attr(feature = "dev", feature(plugin))]
> > +#![cfg_attr(feature = "dev", plugin(clippy(conf_file="../clippy.toml")))]
>
> I created a new feature so we can control if/when CI includes clippy. Do
> you think we should just include it every time unstable is tested?
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/diesel-rs/diesel/pull/305/files/9b4d3b77b7c5c046956b10be79757433abbec0b2#r61047346>
>

sgrif commented on diesel/src/query_dsl/boxed_dsl.rs:
```diff
@@ -57,14 +57,14 @@ impl<'a, T, DB> InternalBoxedDsl<'a, DB> for T where
 ///
 /// [iterator]: https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html
-/// [helper_types]: ../helper_types/index.html
+/// [helpertypes]: ../helper_types/index.html
```
Lol what

killercup replied:
Oh, I tried making sense of the doc_markdown lint which wanted me to put
helper_types in back ticks. I didn't mean to leave that in. (Is there a
clippy issue for that? I should probably create one otherwise)
Sean Griffin <notifications@github.com> schrieb am Mo., 25. Apr. 2016 um
23:24:

> In diesel/src/query_dsl/boxed_dsl.rs
> <https://github.com/diesel-rs/diesel/pull/305#discussion_r60992194>:
>
> >  /// but you might want to hide the return type or have it conditionally change.
> >  /// Boxing can achieve both.
> >  ///
> >  /// [iterator]: https://doc.rust-lang.org/stable/std/iter/trait.Iterator.html
> > -/// [helper_types]: ../helper_types/index.html
> > +/// [helpertypes]: ../helper_types/index.html
>
> Lol what
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/diesel-rs/diesel/pull/305/files/9b4d3b77b7c5c046956b10be79757433abbec0b2#r60992194>
>

killercup replied:
cf. https://github.com/Manishearth/rust-clippy/issues/883

killercup commented:
(Updated to split into two commits and remove change for doc_markdown false positive.)

killercup commented:
I just updated the PR to update clippy and deny warnings _unless_ clippy is used. I also added a clippy 'step' to travis. You can see some of the remaining warnings here: https://travis-ci.org/diesel-rs/diesel/jobs/126730544#L1039

sgrif commented on diesel/src/migrations/migration_error.rs:
```diff
@@ -55,6 +55,7 @@ impl From<io::Error> for MigrationError {
 }
 
 #[derive(Debug, PartialEq)]
+#[cfg_attr(feature = "dev", allow(enum_variant_names))]
```
Do you hold the opinions that the current names are better than what clippy wants, or is this just to avoid the API change?

sgrif replied:
By which I mean I don't feel strongly about this enum, and I'd be fine with changing the names for this one.

killercup replied:
I wasn't sure if it was a breaking change, so I left it as is. I think "QueryError" reads nicely, but if it's only used as "RunMigrationsError::QueryError" I would remove the suffix.

sgrif commented on diesel/src/pg/types/floats/mod.rs:
```diff
@@ -43,12 +43,12 @@ impl Error for InvalidNumericSign {
         let mut bytes = not_none!(bytes);
-        let ndigits = try!(bytes.read_u16::<BigEndian>());
-        let mut digits = Vec::with_capacity(ndigits as usize);
+        let num_digits = try!(bytes.read_u16::<BigEndian>());
```
This is objectively a better name. It's worth noting that I named these local variables for parity with PG internally though, which is what this is based off of. https://github.com/postgres/postgres/blob/6928484bda454f9ab2456d385b2d317f18b6bf1a/src/backend/utils/adt/numeric.c#L822-L825

I don't feel strongly about whether that matters or what this should be named, but I just wanted to mention it. We should go with whichever you think is best.

killercup replied:
Clippy found this because the names `ndigits` and `digits` were to similar, I was actually glad that it did. I had to think about it for a bit to understand what was meant here and I initially thought this was `n` represented as bytes (but there was no `n`). I think `num_*` is still a bit weird (but was just 3 chars to add), I would probably write `digit_count`.

sgrif commented on diesel/src/query_builder/debug.rs:
```diff
@@ -2,7 +2,7 @@ use backend::Debug;
 
 #[doc(hidden)]
-#[derive(Debug)]
+#[derive(Debug, Default)]
```
I wish `Default` had `fn new() -> Self` as an alias for `Default`. Since we can derive this, and the derived fn will be identical to our explicit `new` function. Since this struct is purely internal, I wonder if it's worth just deleting our `new` function entirely and changing the call site to use `default`. It reads slightly worse, but if we're adding the derive we may as well have less code as a result.

killercup replied:
Interesting. I agree, I would remove that `fn new() -> Self`.

I rarely have `fn new() -> Self` (i.e. with no parameter) but often use `fn new(stuff) -> Self` as a constructor (which may internally do stuff like `StructName { field1: stuff, ... StructName::default() }`).

sgrif commented on diesel/Cargo.toml:
```diff
@@ -17,6 +17,7 @@ byteorder = "0.3.*"
 quickcheck = { version = "0.2.25", optional = true }
 chrono = { version = "^0.2.17", optional = true }
 uuid = { version = "^0.2.0", optional = true, features = ["use_std"] }
+clippy = {optional = true, version = "0.0.64"}
```
This version number makes me chuckle.

killercup replied:
We should plan to from that directly to 0.0.128, obviously 😄

killercup commented:
I just rebased this on master and updated to the latest clippy version (maybe that needs a newer nightly -- Travis will tell us).

I've also added some _allow_s for lints which I don't think are useful here ([option_map_unwrap_or_else](https://github.com/Manishearth/rust-clippy/wiki#option_map_unwrap_or_else) and [option_map_unwrap_or](https://github.com/Manishearth/rust-clippy/wiki#option_map_unwrap_or) or for specific parts of the code I didn't want to change. `git grep -i 'feature = "dev", allow'` should show all instances.

killercup commented:
Travis confirms that this clippy version does not compile with the nightly from May 9. @sgrif, I think you mentioned that in the future you wanted to use the nightly from the date the latest stable was released, right? According to http://rusty-dash.com that'd currently be May 27 and the next one will be Jul 8.

I'm okay with downgrading clippy to get this merged, btw.

killercup commented:
I finally updated this PR to a newer clippy version and to fix all the lints (I also enabled a few optional ones). This passes locally with a nightly from 2016-09-29. Once we bump nightly (probably along with the macros 1.1 stuff), I'd like to try to land this.

After that, I'll add clippy to other crates.

sgrif commented:
Sounds good. After Macros 1.1 I'm going to finish composite primary keys and roll a release, we can merge after that

sgrif commented:
Want to rebase so we can merge?

killercup commented:
Rebased, bumped Clippy to version 0.0.103, and allowed/fixed some new lints. Since Clippy fails on current nightly, I made Travis stick to the one from clippy's last release.

Let's see if they manage to release 0.0.104 soon and merge _after that_.

killercup commented:
Rebased, again :D

The most recent nightly is from the 19th (6 days ago), which is kinda unfortunate. Since this is open for 3/4 of a year now, I'll not squash this all right now but wait for one more nightly, and maybe for @sgrif to have quick look.

Ideally, I'd like to add another travis env with a fixed nightly that runs clippy (and is an allowed failure), until we can install clippy with rustup and have a guarantee that it compiles at all times.

killercup commented:
Finally! This is

- adding the latest clippy 0.0.103 as an optional dependency with the `lint` feature
- improving a whole bunch of stuff in the various crates to make clippy real happy
- green with `#[deny(warnings)]`!
- adding a new entry in Travis' build matrix (which compiles quite fast as it is using `no-trans`)

sgrif commented on clippy.toml:
```diff
@@ -0,0 +1,10 @@
+  "GitHub",
+  "IPv4", "IPv6",
+  "JavaScript", "NaN", "OAuth",
+  "SQLite", "PostgreSQL"
```
What's this array actually doing? Are we sure we want to allow this? For example I named everything `Mysql` instead of `MySQL` to be ideomatic. I would assume we want the same to apply to other backends.

killercup replied:
This is only about doc comments, and which CamelCasel-looking things to allow to be used without putting them in back ticks.

sgrif replied:
Ah, I see. We should probably add `MySQL` then. Do we actually need the two lines above this one? `NaN` probably should be in backticks when it appears actually.

sgrif commented on diesel/src/lib.rs:
```diff
@@ -1,9 +1,26 @@
+    option_map_unwrap_or_else, option_map_unwrap_or,
+    match_same_arms, type_complexity,
+))]
+#![cfg_attr(feature = "clippy", warn(
```
This still errors since we have `deny(warnings)` right?

killercup replied:
Indeed. This is to enable additional lints which are allow by default.

sgrif commented on diesel/src/macros/insertable.rs:
```diff
@@ -231,6 +231,7 @@ macro_rules! Insertable_column_expr {
 
 #[cfg(test)]
 #[allow(missing_debug_implementations, missing_copy_implementations)]
+#[cfg_attr(feature = "clippy", allow(result_unwrap_used))]
```
Isn't this redundant with https://github.com/diesel-rs/diesel/pull/305/files#diff-7ab3c70a207718b2fce878a8f925a415R22?

sgrif commented on diesel/src/migrations/migration.rs:
```diff
@@ -101,6 +109,7 @@ fn run_sql_from_file(conn: &SimpleConnection, path: &Path) -> Result<(), RunMigr
 }
 
 #[cfg(test)]
+#[cfg_attr(feature = "clippy", allow(result_unwrap_used))]
```
Isn't this redundant with https://github.com/diesel-rs/diesel/pull/305/files#diff-7ab3c70a207718b2fce878a8f925a415R22?

sgrif commented on diesel/src/migrations/mod.rs:
```diff
@@ -264,6 +263,7 @@ pub fn search_for_migrations_directory(path: &Path) -> Result<PathBuf, Migration
 }
 
 #[cfg(test)]
+#[cfg_attr(feature = "clippy", allow(result_unwrap_used))]
```
Isn't this redundant with https://github.com/diesel-rs/diesel/pull/305/files#diff-7ab3c70a207718b2fce878a8f925a415R22?

sgrif commented on diesel/src/mysql/connection/raw.rs:
```diff
@@ -46,10 +46,10 @@ impl RawConnection {
             ffi::mysql_real_connect(
                 self.0,
-                host.map(|x| x.as_ptr()).unwrap_or(ptr::null_mut()),
+                host.map(CStr::as_ptr).unwrap_or_else(|| ptr::null_mut()),
```
What's the lint that led to the `unwrap_or_else` change? `ptr::null_mut` is a `const fn` so it probably shouldn't be complaining here.

sgrif replied:
https://github.com/Manishearth/rust-clippy/issues/1527

sgrif commented on diesel/src/pg/connection/mod.rs:
```diff
@@ -107,11 +107,12 @@ impl Connection for PgConnection {
 }
 
 impl PgConnection {
+    #[cfg_attr(feature = "clippy", allow(type_complexity))]
```
Heh. I need to refactor this.

sgrif commented on diesel/src/pg/connection/mod.rs:
```diff
@@ -142,7 +143,9 @@ extern "C" fn noop_notice_processor(_: *mut libc::c_void, _message: *const libc:
-    ::std::io::stderr().write(c_str.to_bytes()).unwrap();
+    ::std::io::stderr()
+        .write_all(c_str.to_bytes())
+        .expect("Error writing to `stderr`");
```
The computer is on fire. :)

killercup replied:
Only the screen that shows errors! 🤣

sgrif commented on diesel/src/pg/connection/raw.rs:
```diff
@@ -45,6 +45,7 @@ impl RawConnection {
         RawResult::new(PQexec(self.internal_connection, query), self)
     }
 
+    #[cfg_attr(feature = "clippy", allow(too_many_arguments))]
```
We should probably enable this for the whole file

sgrif commented on diesel/src/pg/connection/result.rs:
```diff
@@ -40,7 +40,8 @@ impl PgResult {
                 "" => 0,
-                _ => count_str.parse().unwrap()
+                _ => count_str.parse()
+                    .expect("Error parsing `rows_affected` as integer value")
```
Now that I'm more familiar with the wire protocol, I know why libpq is giving us this as a c string. And it's really dumb.

sgrif commented on diesel/src/pg/connection/stmt/mod.rs:
```diff
@@ -24,6 +24,7 @@ pub enum Query {
 }
 
 impl Query {
+    #[cfg_attr(feature = "clippy", allow(ptr_arg))]
```
We should probably enable this for the whole module.

sgrif commented on diesel/src/pg/connection/stmt/mod.rs:
```diff
@@ -34,8 +35,10 @@ impl Query {
                  .unwrap_or(ptr::null()))
```
Wait did clippy not break on this one?

killercup replied:
Huh. Does it have a special case for null but not null_mut?

sgrif commented on diesel/src/pg/types/floats/mod.rs:
```diff
@@ -41,12 +41,12 @@ impl Error for InvalidNumericSign {
 impl FromSql<types::Numeric, Pg> for PgNumeric {
     fn from_sql(bytes: Option<&[u8]>) -> Result<Self, Box<Error+Send+Sync>> {
         let mut bytes = not_none!(bytes);
-        let ndigits = try!(bytes.read_u16::<NetworkEndian>());
```
I named this variable for parity with the pg code which is the "documentation" for this structure. I don't feel strongly on whether we need to keep it or not.

killercup replied:
That's a good reason. I'll allow the lint here (and add a link to the pg code)

sgrif commented on diesel/src/connection/transaction_manager.rs:
```diff
@@ -31,19 +31,25 @@ pub trait TransactionManager<Conn: Connection> {
-impl AnsiTransactionManager {
-    pub fn new() -> Self {
+impl Default for AnsiTransactionManager {
+    fn default() -> Self {
```
I think we can derive this.

sgrif commented on diesel/src/sqlite/connection/mod.rs:
```diff
@@ -182,6 +182,7 @@ fn error_message(err_code: libc::c_int) -> &'static str {
 }
 
 #[cfg(test)]
+#[cfg_attr(feature = "clippy", allow(result_unwrap_used))]
```
Isn't this redundant with the line in `lib.rs`?

sgrif commented on diesel/src/sqlite/query_builder/mod.rs:
```diff
@@ -10,14 +10,20 @@ pub struct SqliteQueryBuilder {
-impl SqliteQueryBuilder {
-    pub fn new() -> Self {
+impl Default for SqliteQueryBuilder {
+    fn default() -> Self {
```
I think we can derive this

sgrif commented on diesel/src/test_helpers.rs:
```diff
@@ -1,3 +1,4 @@
+#![cfg_attr(feature = "clippy", allow(result_unwrap_used))]
```
Isn't this redundant with the line in `lib.rs`?

sgrif commented on diesel/src/types/impls/option.rs:
```diff
@@ -110,6 +110,7 @@ use pg::Pg;
 
 #[test]
 #[cfg(feature = "postgres")]
+#[cfg_attr(feature = "clippy", allow(string_lit_as_bytes))]
```
Eh. We may as well just change that line to `b"string lit"`. Also there's a typo in `expected` on that line, too

sgrif commented on diesel/src/types/impls/tuples.rs:
```diff
@@ -19,8 +19,8 @@ macro_rules! tuple_impls {
             impl<$($T),+, DB> HasSqlType<($($T,)+)> for DB where
                 $(DB: HasSqlType<$T>),+,
                 DB: Backend,
-            {
```
I don't think you meant to make this change.

sgrif commented on diesel/src/types/impls/tuples.rs:
```diff
@@ -71,6 +71,7 @@ macro_rules! tuple_impls {
                 type SqlType = ($(<$T as Expression>::SqlType,)+);
             }
 
+            #[cfg_attr(feature = "clippy", allow(eq_op))] // Clippy doesn't like the trivial case for 1-tuples
```
Should we enable this for the whole file?

sgrif commented on diesel_cli/Cargo.toml:
```diff
@@ -18,13 +18,15 @@ clap = "2.20.3"
 diesel = { version = "0.10.0", default-features = false }
 dotenv = "0.8.0"
 diesel_infer_schema = { version = "0.1.0", default-features = false, path = "../diesel_infer_schema" }
+clippy = {optional = true, version = "0.0.113"}
```
Spaces?

killercup replied:
[Sure](https://i.ytimg.com/vi/DHibe7XxZAQ/maxresdefault.jpg)

sgrif commented on diesel_cli/src/main.rs:
```diff
@@ -1,5 +1,20 @@
+#![deny(warnings, missing_copy_implementations)]
+
+// Clippy lints
+#![cfg_attr(feature = "clippy", allow(unstable_features))]
```
Should we stick this in a file that we can `include!` in all the crates?

killercup replied:
Good idea. Where should I put that?

killercup replied:
Actually, if we put that in the root directory, how can we make sure the `include!`d file gets, well, included in the `cargo package` artifact?

sgrif replied:
Not sure. Let's just try it and see what happens. I'd put it in the project root directory and call it `make_crate_pedantic.rs`

sgrif replied:
Though CLI should probably continue to allow things in release builds so that random lints added in future versions of Rust don't cause the binary to stop compiling.

sgrif commented:
Can you add the `lint` feature to the integration test lines in `bin/test`, too? (Maybe hold off for a few on rebasing I'm about to push up something that changes that file for MySQL)

sgrif commented on diesel/Cargo.toml:
```diff
@@ -21,6 +21,7 @@ quickcheck = { version = "0.3.1", optional = true }
 serde_json = { version = ">=0.8.0, <0.10.0", optional = true }
 uuid = { version = ">=0.2.0, <0.5.0", optional = true, features = ["use_std"] }
 url = { version = "1.4.0", optional = true }
+clippy = {optional = true, version = "0.0.114"}
```
Does this need to be locked down to `=0.0.114`?

killercup replied:
Hm, ^0.0.X should be equivalent to =0.0.X IIRC. But better safe than sorry! 😅

killercup commented:
🎉
